README.md
# hot-shots-unix

This is a fork of [hot-shots](https://github.com/brightcove/hot-shots) that adds support for UDS.

---

A Node.js client for [Etsy](http://etsy.com)'s [StatsD](https://github.com/etsy/statsd) server, Datadog's [DogStatsD](http://docs.datadoghq.com/guides/dogstatsd/) server, and [InfluxDB's](http://influxdb.com) [Telegraf](https://github.com/influxdb/telegraf) StatsD server.

This project was originally a fork off of [node-statsd](https://github.com/sivy/node-statsd).  This project
includes all changes in the latest node-statsd and many additional changes, including:
* TypeScript types
* Telegraf support
* events
* child clients
* tcp protocol support
* mock mode
* asyncTimer
* much more, including many bug fixes

hot-shots supports Node 6.x and higher.

[![CircleCI](https://circleci.com/gh/msindwan/hot-shots.svg?style=svg)](https://circleci.com/gh/msindwan/hot-shots)

## Migrating from node-statsd

You should only need to do one thing: change node-statsd to hot-shots in all requires.

You may also want to use the Datadog events support in here instead of other libraries.  You can also check the detailed [change log](https://github.com/brightcove/hot-shots/blob/master/CHANGES.md) for what has changed since the last release of node-statsd.

## Usage

All initialization parameters are optional.

Parameters (specified as one object passed into hot-shots):
* `host`:        The host to send stats to `default: localhost`
* `port`:        The port to send stats to `default: 8125`
* `path`:        The path to the unix domain socket if using `unix_dgram` protocol
* `prefix`:      What to prefix each stat name with `default: ''`
* `suffix`:      What to suffix each stat name with `default: ''`
* `globalize`:   Expose this StatsD instance globally? `default: false`
* `cacheDns`:    Cache the initial dns lookup to *host* `default: false`
* `mock`:        Create a mock StatsD instance, sending no stats to the server and allowing data to be read from mockBuffer? `default: false`
* `globalTags`:  Tags that will be added to every metric. Can be either an object or list of tags. `default: {}`
* `maxBufferSize`: If larger than 0,  metrics will be buffered and only sent when the string length is greater than the size. `default: 0`
* `bufferFlushInterval`: If buffering is in use, this is the time in ms to always flush any buffered metrics. `default: 1000`
* `telegraf`:    Use Telegraf's StatsD line protocol, which is slightly different than the rest `default: false`
* `sampleRate`:    Sends only a sample of data to StatsD for all StatsD methods.  Can be overridden at the method level. `default: 1`
* `errorHandler`: A function with one argument. It is called to handle various errors. `default: none`, errors are thrown/logger to console
* `useDefaultRoute`: Use the default interface on a Linux system. Useful when running in containers
* `protocol`: Use `tcp` option for TCP protocol. Defaults to UDP otherwise

All StatsD methods other than event and close have the same API:
* `name`:       Stat name `required`
* `value`:      Stat value `required except in increment/decrement where it defaults to 1/-1 respectively`
* `sampleRate`: Sends only a sample of data to StatsD `default: 1`
* `tags`:       The tags to add to metrics. Can be either an object `{ tag: "value"}` or an array of tags. `default: []`
* `callback`:   The callback to execute once the metric has been sent or buffered

If an array is specified as the `name` parameter each item in that array will be sent along with the specified value.

The close method has the following API:

* `callback`:   The callback to execute once close is complete.  All other calls to statsd will fail once this is called.

The event method has the following API:

* `title`:       Event title `required`
* `text`:        Event description `default is title`
* `options`:     Options for the event
  * `date_happened`    Assign a timestamp to the event `default is now`
  * `hostname`         Assign a hostname to the event.
  * `aggregation_key`  Assign an aggregation key to the event, to group it with some others.
  * `priority`         Can be ‘normal’ or ‘low’ `default: normal`
  * `source_type_name` Assign a source type to the event.
  * `alert_type`       Can be ‘error’, ‘warning’, ‘info’ or ‘success’ `default: info`
* `tags`:       The tags to add to metrics. Can be either an object `{ tag: "value"}` or an array of tags. `default: []`
* `callback`:   The callback to execute once the metric has been sent.

The check method has the following API:

* `name`:        Check name `required`
* `status`:      Check status `required`
* `options`:     Options for the check
  * `date_happened`    Assign a timestamp to the check `default is now`
  * `hostname`         Assign a hostname to the check.
  * `message`          Assign a message to the check.
* `tags`:       The tags to add to metrics. Can be either an object `{ tag: "value"}` or an array of tags. `default: []`
* `callback`:   The callback to execute once the metric has been sent.

```javascript
  var StatsD = require('hot-shots'),
      client = new StatsD({ port: 8020, globalTags: { env: process.env.NODE_ENV }, errorHandler: errorHandler });

  // Timing: sends a timing command with the specified milliseconds
  client.timing('response_time', 42);

  // Timer: Returns a function that you call to record how long the first
  // parameter takes to execute (in milliseconds) and then sends that value
  // using 'client.timing'.
  // The parameters after the first one (in this case 'fn')
  // match those in 'client.timing'.
  var fn = function(a, b) { return a + b };
  client.timer(fn, 'fn_execution_time')(2, 2);

  // Async timer: Similar to timer above, but you instead pass in a funtion
  // that returns a Promise.  And then it returns a Promise that will record the timing.
  var fn = function () { return new Promise(function (resolve, reject) { setTimeout(resolve, n); }); };
  var instrumented = statsd.asyncTimer(fn, 'fn_execution_time');
  instrumented().then(function() {
    console.log('Code run and metric sent');
  });

  // Increment: Increments a stat by a value (default is 1)
  client.increment('my_counter');

  // Decrement: Decrements a stat by a value (default is -1)
  client.decrement('my_counter');

  // Histogram: send data for histogram stat (DataDog and Telegraf only)
  client.histogram('my_histogram', 42);

  // Distribution: Tracks the statistical distribution of a set of values across your infrastructure. (DataDog v6)
  client.distribution('my_distribution', 42);

  // Gauge: Gauge a stat by a specified amount
  client.gauge('my_gauge', 123.45);

  // Set: Counts unique occurrences of a stat (alias of unique)
  client.set('my_unique', 'foobar');
  client.unique('my_unique', 'foobarbaz');

  // Event: sends the titled event (DataDog only)
  client.event('my_title', 'description');

  // Check: sends a service check (DataDog only)
  client.check('service.up', client.CHECKS.OK, { hostname: 'host-1' }, ['foo', 'bar'])

  // Incrementing multiple items
  client.increment(['these', 'are', 'different', 'stats']);

  // Incrementing with tags
  client.increment('my_counter', ['foo', 'bar']);

  // Sampling, this will sample 25% of the time the StatsD Daemon will compensate for sampling
  client.increment('my_counter', 1, 0.25);

  // Tags, this will add user-defined tags to the data (DataDog and Telegraf only)
  client.histogram('my_histogram', 42, ['foo', 'bar']);

  // Using the callback.  This is the same format for the callback
  // with all non-close calls
  client.set(['foo', 'bar'], 42, function(error, bytes){
    //this only gets called once after all messages have been sent
    if(error){
      console.error('Oh noes! There was an error:', error);
    } else {
      console.log('Successfully sent', bytes, 'bytes');
    }
  });

  // Sampling, tags and callback are optional and could be used in any combination (DataDog and Telegraf only)
  client.histogram('my_histogram', 42, 0.25); // 25% Sample Rate
  client.histogram('my_histogram', 42, { tag: 'value'}]); // User-defined tag
  client.histogram('my_histogram', 42, ['tag:value']); // Tags as an array
  client.histogram('my_histogram', 42, next); // Callback
  client.histogram('my_histogram', 42, 0.25, ['tag']);
  client.histogram('my_histogram', 42, 0.25, next);
  client.histogram('my_histogram', 42, { tag: 'value'}, next);
  client.histogram('my_histogram', 42, 0.25, { tag: 'value'}, next);

  // Use a child client to add more context to the client.
  // Clients can be nested.
  var childClient = client.childClient({
    prefix: 'additionalPrefix.',
    suffix: '.additionalSuffix',
    globalTags: { globalTag1: 'forAllMetricsFromChildClient'}
  });
  childClient.increment('my_counter_with_more_tags');

  // Close statsd.  This will ensure all stats are sent and stop statsd
  // from doing anything more.
  client.close(function(err) {
    console.log('The close did not work quite right: ', err);
  });
```

## DogStatsD and Telegraf functionality

Some of the functionality mentioned above is specific to DogStatsD or Telegraf.  They will not do anything if you are using the regular statsd client.
* globalTags parameter- DogStatsD or Telegraf
* tags parameter- DogStatsD or Telegraf.
* telegraf parameter- Telegraf
* histogram method- DogStatsD or Telegraf
* event method- DogStatsD
* check method- DogStatsD

## Errors

As usual, callbacks will have an error as their first parameter.  You can have an error in both the message and close callbacks.

If the optional callback is not given, an error is thrown in some cases. An error will only be explicitly thrown when there is a missing callback or if it is some potential configuration issue to be fixed.

If you would like to ensure all errors are caught, specify an `errorHandler` in your root client. This will catch errors in socket setup, sending of messages, and closing of the socket. If you specify an errorHandler and a callback, the callback will take precedence.

```javascript
// Using errorHandler
var client = new StatsD({
  errorHandler: function (error) {
    console.log("Socket errors caught here: ", error);
  }
})
```

## Submitting changes

Thanks for considering making any updates to this project!  Here are the steps to take in your fork:

1. Run "npm install"
2. Add your changes in your fork as well as any new tests needed
3. Run "npm test"
4. Run "npm benchmark" and observe any anomalies
5. Update README.md with any needed documentation
6. If you have made any API changes, update types.d.ts
7. Push your changes and create the PR

When you've done all this we're happy to try to get this merged in right away.

## Name

Why is this project named hot-shots?  Because:

1. It's impossible to find another statsd name on npm
2. It's the name of a dumb movie
3. No good reason

## License

hot-shots is licensed under the MIT license.


CHANGES.md
CHANGELOG
=========

## 7.0.1 (2019-3-12)
* @msindwan remove postinstall script

## 7.0.0 (2019-3-08)
* @msindwan Add UDS support
* @msindwan Add unit tests and switch to Jest
* @msindwan Use es6 classes

* @JamesMGreene Fix syntax in README example

## 6.1.1 (2019-1-8)
* @bdeitte Fix errorHandler to only happen again on errors
* @Ithildir Readme fixes

## 6.1.0 (2019-1-5)
* @bdeitte Ensure close() call always sends data before closing
* @bdeitte Recommend errorHandler over client.socket.on() for handling
errors
* @mbellerose Fix the timer function type definition

## 6.0.1 (2018-12-17)
* @msmnc Fix regression when tag value is a number
* @bdeitte Make non-options in constructor more deprecated

## 6.0.0 (2018-12-15)
@bdeitte Major upgrade to the codebase to be more modern,
overhaul tests, and many small tweaks.  Most of this is internal to
the project, but there are a few changes to note for everyone:
* Now requires Node 6 or above
* Update close() to handle errors better, not doubling up in error
messages and not leaving uncaught errors

Everything else done here should be internal facing.  Those changes
include:
* Use "lebab" to ES6-ify the project
* Switch from jshint and eslint and make syntax updates based on this
* Remove a lot of duplication in tests and many small fixups in tests
* Start using Mocha 4
* Stop using index.js for testing
* Start using the code coverage report as part of the build
* Remove the ignoring of errors on close of tests, and tear down tests in general better
* Stop using "new Buffer", that is deprecated, and use Buffer.from() instead

## 5.9.2 (2018-11-10)
* @stieg Add mockBuffer to types

## 5.9.1 (2018-9-18)
* @etaoins Add asyncTimer types
* @blimmer: Add increment doc snippet

## 5.9.0 (2018-7-27)
* @chrismatheson: Fix timer to have duration in microseconds (was nanoseconds)
* @chrismatheson: Add asyncTimer functionality

## 5.8.0 (2018-7-17)
* @michalholasek Clean up code formatting and split up tests
* @michalholasek Add tcp protocol support
* @remie Add tcp protocol support

## 5.7.0 (2018-7-4)
* @Willyham Add support for recording buffers in mock mode

## 5.6.3 (2018-6-20)
* @singerb correct close() type definition

## 5.6.2 (2018-6-15)
* @mjesuele Fix time in timer

## 5.6.1 (2018-6-4)
* @MattySheikh Typescript: add socket type for StatsD class

## 5.6.0 (2018-6-3)
* @drewen TypeScript: add overload types for stats functions

## 5.5.1 (2018-5-30)
* @emou Typescript declaration for the 'timer' method

## 5.5.0 (2018-5-30)
* @drewen Split up single file, add code coverage capabilities

## 5.4.1 (2018-5-12)
* @jasonsack Fixups for new useDefaultRoute option
* @bdeitte Test against more modern set of Node versions in Travis

## 5.4.0 (2018-4-26)
* @RobGraham Added `distribution()` support for DataDog v6

## 5.3.0 (2018-4-3)
* @tanelso2 Added support for using default route on Linux

## 5.2.0 (2018-2-28)
* @ericapisani Add timer decorator function

## 5.1.0 (2018-2-14)
* @lautis Pass key-value tags as objects

## 5.0.1 (2018-2-2)
* @punya-asapp Add childClient to TypeScript types

## 5.0.0 (2017-11-9)
* @jgwmaxwell TypeScript typings, resolving the default export issue and missing options from last time.  This is being marked as a major release, in caution given the revert last time, but it is not actually known to cause any backwards-compatible issues.

## 4.8.0 (2017-10-31)
* @Jiggmin concat prefix and suffix in check function
* @Jiggmin commit package-lock.json

## 4.7.1 (2017-10-31)
* @Jiggmin Add backwards compatibility for global_tags

## 4.7.0 (2017-9-21)
* @bdeitte Revert TypeScript typings, which ended up not being semver minor

## 4.6.0 (2017-9-19)
* @jgwmaxwell TypeScript typings

## 4.5.0 (2017-5-4)
* @jsocol Support default value with tags in increment

## 4.4.0 (2017-3-23)
* @RijulB Global sample rate

## 4.3.1 (2016-11-7)
* @RandomSeeded Fix callbacks not being triggered when using buffers

## 4.3.0 (2016-9-30)
* @ggoodman Allow socket errors to be handled with errorHandler

## 4.2.0 (2016-8-3)
* @mhahn Add support for DataDog service checks

## 4.1.1 (2016-5-22)
* @ash2k date_happened should be seconds, not milliseconds

## 4.1.0 (2016-5-8)
* @ash2k Support multiline text in DataDog events

## 4.0.0 (2016-5-7)
* @ash2k Provided tags, including `childClient()` tags, override global tags with same names.

## 3.1.0 (2016-5-3)
* @ash2k Support a client-wide error handler used in case no callback is provided and to handle various exceptions.

## 3.0.1 (2016-4-28)
* @bdeitte Add 'use strict' to files and make changes needed for this.

## 3.0.0 (2016-4-27)
* @ash2k Method to create child clients.  (This is not a backwards-incompatible change but is rather large.)
* @ash2k Shrink npm package a bit more

## 2.4.0 (2016-2-26)
* @arlolra Shrink npm package
* @arlolra/@bdeitte Move DNS errors when caching them to send() and use callback when possible
* @bdeitte Use callback for Telegraf error when possible

## 2.3.1 (2016-2-3)
* @Pchelolo Ensure messages not larger then maxBufferSize

## 2.3.0 (2016-1-17)
* @bdeitte Fix increment(name, 0) to send a 0 count instead of 1
* @bdeitte Flush the queue when needed on close()

## 2.2.0 (2016-1-10)
* @bdeitte Document and expand on close API
* @bdeitte Catch more error cases for callbacks

## 2.1.2 (2015-12-9)
* @bdeitte Even more doc updates
* @mmoulton Fix multiple tags with Telegraf

## 2.1.1 (2015-12-9)
* @bdeitte Doc updates

## 2.1.0 (2015-12-9)
* @mmoulton Add options.telegraf to enable support for Telegraf's StatsD line protocol format
* @mmoulton Ensure message callback is sent in buffered case, even when we just buffer.

## 2.0.0 (2015-10-22)
* @jjofseattle Add options.maxBufferSize and options.bufferFlushInterval
* @bdeitte Change options.global_tags to options.globalTags for consistency

## 1.0.2 (2015-09-25)
* @ainsleyc Thrown error when cacheDNS flag fails to resolve DNS name

## 1.0.1 (2015-09-24)
* @bdeitte Add the event API used by DogStatsD

tests/unit/statsd.unit.test.js
const StatsD = require('../../index');

jest.mock(
  'unix-dgram',
  () => {
    return {
      createSocket() {
        return {
          connect: jest.fn(),
          send: jest.fn(),
          close: jest.fn(),
          on: jest.fn(),
        };
      },
    };
  },
  { virtual: true }
);

describe('statsd with unix_dgram protocol', () => {
  const originalMathRandom = global.Math.random;
  let client;

  beforeAll(() => {
    global.Math.random = () => 0.3;
  });

  afterAll(() => {
    global.Math.random = originalMathRandom;
  });

  afterEach(() => {
    jest.clearAllMocks();
    if (client) {
      client.close();
      client = null;
    }
  });

  test('mock disables emitting stats', () => {
    client = new StatsD({ maxBufferSize: 0, protocol: 'unix_dgram', mock: true });
    expect(client.mock).toEqual(true);

    client.increment('my.stat');
    expect(client.socket.send).toHaveBeenCalledTimes(0);
  });

  test('setting the sampling rate to 0 disables emitting stats', () => {
    client = new StatsD({ maxBufferSize: 0, protocol: 'unix_dgram', sampleRate: 0, mock: false });

    client.increment('my.stat');
    expect(client.socket.send).toHaveBeenCalledTimes(0);
  });

  test('emitted stats include sampling rate if < 1', () => {
    client = new StatsD({
      maxBufferSize: 0,
      protocol: 'unix_dgram',
      sampleRate: 0.5,
      mock: false,
      path: '/var/run/datadog/dsd.socket'
    });

    client.increment('my.stat');
    const buffer = Buffer.from('my.stat:1|c|@0.5');

    expect(client.socket.send).toHaveBeenCalledTimes(1);
    expect(client.socket.send).toHaveBeenCalledWith(
      buffer,
      expect.any(Function)
    );
  });

  test('emitted stats include custom tags', () => {
    client = new StatsD({
      maxBufferSize: 0,
      mock: false,
      protocol: 'unix_dgram',
      path: '/var/run/datadog/dsd.socket'
    });

    client.increment('my.stat', 5, ['a:tag', 'b:otherTag']);
    const buffer = Buffer.from('my.stat:5|c|#a:tag,b:otherTag');

    expect(client.socket.send).toHaveBeenCalledTimes(1);
    expect(client.socket.send).toHaveBeenCalledWith(
      buffer,
      expect.any(Function)
    );
  });

  test('emitted stats include global prefix', () => {
    client = new StatsD({
      maxBufferSize: 0,
      mock: false,
      prefix: 'myPrefix.',
      protocol: 'unix_dgram',
      path: '/var/run/datadog/dsd.socket'
    });

    client.increment('my.stat', 5, ['a:tag', 'b:otherTag']);
    const buffer = Buffer.from('myPrefix.my.stat:5|c|#a:tag,b:otherTag');

    expect(client.socket.send).toHaveBeenCalledTimes(1);
    expect(client.socket.send).toHaveBeenCalledWith(
      buffer,
      expect.any(Function)
    );
  });

  test('stats are emitted if the in-memory buffer is disabled', () => {
    client = new StatsD({
      maxBufferSize: 0,
      mock: false,
      protocol: 'unix_dgram',
      path: '/var/run/datadog/dsd.socket'
    });
    client.increment('inc1.stat');
    client.increment('inc2.stat');
    client.decrement('dec1.stat');
    client.decrement('dec2.stat');

    expect(client.socket.send).toHaveBeenCalledTimes(4);

    [
      Buffer.from('inc1.stat:1|c'),
      Buffer.from('inc2.stat:1|c'),
      Buffer.from('dec1.stat:-1|c'),
      Buffer.from('dec2.stat:-1|c'),
    ].forEach((buffer, index) => {
      expect(client.socket.send).toHaveBeenNthCalledWith(
        index + 1,
        buffer,
        expect.any(Function)
      );
    });
  });

  test('stats are emitted if the in-memory buffer is enabled', done => {
    client = new StatsD({
      maxBufferSize: 150,
      flushInterval: 500,
      mock: false,
      protocol: 'unix_dgram',
      path: '/var/run/datadog/dsd.socket'
    });
    client.increment('inc1.stat');
    client.increment('inc2.stat');
    client.decrement('dec1.stat');
    client.decrement('dec2.stat');

    setTimeout(() => {
      const buffer = Buffer.from('inc1.stat:1|c\ninc2.stat:1|c\ndec1.stat:-1|c\ndec2.stat:-1|c\n');
      expect(client.socket.send).toHaveBeenCalledTimes(1);
      expect(client.socket.send).toHaveBeenCalledWith(
        buffer,
        expect.any(Function)
      );
      done();
    }, 1000);
  });

  test('stats are emitted if the in-memory buffer is full', () => {
    client = new StatsD({
      maxBufferSize: 30,
      flushInterval: 5000,
      mock: false,
      protocol: 'unix_dgram',
      path: '/var/run/datadog/dsd.socket'
    });
    client.increment('inc1.stat');
    client.increment('inc2.stat');
    client.increment('inc3.stat');

    const buffer = Buffer.from('inc1.stat:1|c\ninc2.stat:1|c\n');
    expect(client.socket.send).toHaveBeenCalledTimes(1);
    expect(client.socket.send).toHaveBeenCalledWith(
      buffer,
      expect.any(Function)
    );
  });

  test('stats are not emitted if the in-memory buffer is empty', done => {
    client = new StatsD({
      maxBufferSize: 30,
      flushInterval: 100,
      mock: false,
      protocol: 'unix_dgram',
      path: '/var/run/datadog/dsd.socket'
    });
    setTimeout(() => {
      expect(client.socket.send).toHaveBeenCalledTimes(0);
      done();
    }, 200);
  });

  test('closing the client emits anything remaning in the buffer', done => {
    client = new StatsD({
      maxBufferSize: 150,
      flushInterval: 500,
      mock: false,
      protocol: 'unix_dgram',
      path: '/var/run/datadog/dsd.socket'
    });
    client.increment('inc1.stat');
    client.close();

    expect(client.socket.send).toHaveBeenCalledTimes(1);
    expect(client.socket.send).toHaveBeenCalledTimes(1);
    setTimeout(() => {
      expect(client.socket.send).toHaveBeenCalledTimes(1);
      done();
    }, 1000);
  });
});


lib/transport.js
const unixDgram = require('unix-dgram');
const helpers = require('./helpers');
const dgram = require('dgram');
const dns = require('dns');
const net = require('net');

/**
 * @description Creates a socket with the arguments provided
 * @param {object} args
 * @param {string} host - The host to use for the socket
 * @param {number} port - The port to use for the socket
 * @param {number} port - The port to use for the socket
 * @throws Error if it fails to create the socket
 */
function createSocket(args) {
  let socket;

  if (args.protocol === 'tcp') {
    socket = net.connect(args.port, args.host);
    socket.setKeepAlive(true);
  } else if (args.protocol === 'unix_dgram') {
    socket = unixDgram.createSocket('unix_dgram');
    socket.connect(args.path);
  } else {
    socket = dgram.createSocket('udp4');
  }

  return socket;
}

/**
 * StatsD Transport
 * @description Base class for transporting StatsD Messages
 * @throws Error if it fails to create the socket
 * NOTE: Adding new parameters to the constructor is deprecated- please use the
 * constructor as one options object.
 */
class Transport {
  constructor(
    host,
    port,
    prefix,
    suffix,
    globalize,
    cacheDns,
    mock,
    globalTags,
    maxBufferSize,
    bufferFlushInterval,
    telegraf,
    sampleRate,
    protocol
  ) {
    let options = host || {};
    const self = this;

    // Adding options below is DEPRECATED.  Use the options object instead.
    if (arguments.length > 1 || typeof(host) === 'string') {
      options = {
        host        : host,
        port        : port,
        prefix      : prefix,
        suffix      : suffix,
        globalize   : globalize,
        cacheDns    : cacheDns,
        mock        : mock === true,
        globalTags  : globalTags,
        maxBufferSize : maxBufferSize,
        bufferFlushInterval: bufferFlushInterval,
        telegraf    : telegraf,
        sampleRate  : sampleRate,
        protocol    : protocol
      };
    }

    // hidden global_tags option for backwards compatibility
    options.globalTags = options.globalTags || options.global_tags;

    this.protocol    = (options.protocol && options.protocol.toLowerCase());
    this.host        = options.host || 'localhost';
    this.port        = options.port || 8125;
    this.path        = options.path || '';
    this.prefix      = options.prefix || '';
    this.suffix      = options.suffix || '';
    this.socket      = options.isChild ? options.socket : createSocket({
      host: this.host,
      path: this.path,
      port: this.port,
      protocol: this.protocol
    });
    this.mock        = options.mock === true;
    this.globalTags  = typeof options.globalTags === 'object' ?
        helpers.formatTags(options.globalTags, options.telegraf) : [];
    this.telegraf    = options.telegraf || false;
    this.maxBufferSize = options.maxBufferSize || 0;
    this.sampleRate  = typeof options.sampleRate === 'number' ? options.sampleRate : 1;
    this.bufferFlushInterval = options.bufferFlushInterval || 1000;
    this.bufferHolder = options.isChild ? options.bufferHolder : { buffer: '' };
    this.errorHandler = options.errorHandler;

    // If we're mocking the client, create a buffer to record the outgoing calls.
    if (this.mock) {
      this.mockBuffer = [];
    }

    // We only want a single flush event per parent and all its child clients
    if (!options.isChild && this.maxBufferSize > 0) {
      this.intervalHandle = setInterval(this.onBufferFlushInterval.bind(this), this.bufferFlushInterval);
    }

    if (options.isChild) {
      if (options.dnsError) {
        this.dnsError = options.dnsError;
      }
    } else if (options.cacheDns === true) {
      dns.lookup(options.host, (err, address) => {
        if (err === null) {
          self.host = address;
        } else {
          self.dnsError = err;
        }
      });
    }

    if (!options.isChild && options.errorHandler) {
      this.socket.on('error', options.errorHandler);
    }

    if (options.globalize) {
      global.statsd = this;
    }

    if (options.useDefaultRoute) {
      const defaultRoute = helpers.getDefaultRoute();
      if (defaultRoute) {
        this.host = defaultRoute;
      }
    }

    this.messagesInFlight = 0;
    this.CHECKS = {
      OK: 0,
      WARNING: 1,
      CRITICAL: 2,
      UNKNOWN: 3,
    };
  }

  /**
   * Checks if stats is an array and sends all stats calling back once all have sent
   * @param stat {String|Array} The stat(s) to send
   * @param value The value to send
   * @param type The type of the metric
   * @param sampleRate {Number=} The Number of times to sample (0 to 1). Optional.
   * @param tags {Array=} The Array of tags to add to metrics. Optional.
   * @param callback {Function=} Callback when message is done being delivered. Optional.
   */
  sendAll(stat, value, type, sampleRate, tags, callback) {
    let completed = 0;
    let calledback = false;
    let sentBytes = 0;
    const self = this;

    if (sampleRate && typeof sampleRate !== 'number') {
      callback = tags;
      tags = sampleRate;
      sampleRate = undefined;
    }

    if (tags && typeof tags !== 'object') {
      callback = tags;
      tags = undefined;
    }

    /**
     * Gets called once for each callback, when all callbacks return we will
     * call back from the function
     * @private
     */
    function onSend(error, bytes) {
      completed += 1;
      if (calledback) {
        return;
      }

      if (error) {
        if (typeof callback === 'function') {
          calledback = true;
          callback(error);
        } else if (self.errorHandler) {
          calledback = true;
          self.errorHandler(error);
        }
        return;
      }

      if (bytes) {
        sentBytes += bytes;
      }

      if (completed === stat.length && typeof callback === 'function') {
        callback(null, sentBytes);
      }
    }

    if (Array.isArray(stat)) {
      stat.forEach(item => {
        self.sendStat(item, value, type, sampleRate, tags, onSend);
      });
    } else {
      this.sendStat(stat, value, type, sampleRate, tags, callback);
    }
  }

  /**
   * Sends a stat across the wire
   * @param stat {String|Array} The stat(s) to send
   * @param value The value to send
   * @param type {String} The type of message to send to statsd
   * @param sampleRate {Number} The Number of times to sample (0 to 1)
   * @param tags {Array} The Array of tags to add to metrics
   * @param callback {Function=} Callback when message is done being delivered. Optional.
   */
  sendStat(stat, value, type, sampleRate, tags, callback) {
    let message = `${this.prefix + stat + this.suffix}:${value}|${type}`;
    sampleRate = typeof sampleRate === 'number' ? sampleRate : this.sampleRate;
    if (sampleRate < 1) {
      if (Math.random() < sampleRate) {
        message += `|@${sampleRate}`;
      } else {
        // don't want to send if we don't meet the sample ratio
        return callback ? callback() : undefined;
      }
    }
    this.send(message, tags, callback);
  }

  /**
   * Send a stat or event across the wire
   * @param message {String} The constructed message without tags
   * @param tags {Array} The tags to include (along with global tags). Optional.
   * @param callback {Function=} Callback when message is done being delivered (only if maxBufferSize == 0). Optional.
   */
  send(message, tags, callback) {
    let mergedTags = this.globalTags;
    if (tags && typeof tags === 'object') {
      mergedTags = helpers.overrideTags(mergedTags, tags, this.telegraf);
    }
    if (mergedTags.length > 0) {
      if (this.telegraf) {
        message = message.split(':');
        message = `${message[0]},${mergedTags.join(',').replace(/:/g, '=')}:${message.slice(1).join(':')}`;
      } else {
        message += `|#${mergedTags.join(',')}`;
      }
    }

    this._send(message, callback);
  }

  /**
   * Send a stat or event across the wire
   * @param message {String} The constructed message without tags
   * @param callback {Function=} Callback when message is done being delivered (only if maxBufferSize == 0). Optional.
   */
  _send(message, callback) {
    // we may have a cached error rather than a cached lookup, so
    // throw it on
    if (this.dnsError) {
      if (callback) {
        return callback(this.dnsError);
      } else if (this.errorHandler) {
        return this.errorHandler(this.dnsError);
      }
      throw this.dnsError;
    }

    // Only send this stat if we're not a mock Client.
    if (!this.mock) {
      if (this.maxBufferSize === 0) {
        this.sendMessage(message, callback);
      } else {
        this.enqueue(message, callback);
      }
    } else {
      this.mockBuffer.push(message);
      if (typeof callback === 'function') {
        callback(null, 0);
      }
    }
  }

  /**
   * Add the message to the buffer and flush the buffer if needed
   *
   * @param message {String} The constructed message without tags
   */
  enqueue(message, callback) {
    message += '\n';

    if (this.bufferHolder.buffer.length + message.length > this.maxBufferSize) {
      this.flushQueue(callback);
      this.bufferHolder.buffer += message;
    }
    else {
      this.bufferHolder.buffer += message;
      if (callback) {
        callback(null);
      }
    }
  }

  /**
   * Flush the buffer, sending on the messages
   */
  flushQueue(callback) {
    this.sendMessage(this.bufferHolder.buffer, callback);
    this.bufferHolder.buffer = '';
  }

  /**
   * Send on the message through the socket
   *
   * @param message {String} The constructed message without tags
   * @param callback {Function=} Callback when message is done being delivered. Optional.
   */
  sendMessage(message, callback) {
    // don't waste the time if we aren't sending anything
    if (message === '') {
      if (callback) {
        callback(null);
      }
      return;
    }

    if (this.protocol === 'tcp' && message.lastIndexOf('\n') !== message.length - 1) {
      message += '\n';
    }

    const handleCallback = (err) => {
      this.messagesInFlight--;
      const errFormatted = err ? new Error(`Error sending hot-shots message: ${err}`) : null;
      if (errFormatted) {
        errFormatted.code = err.code;
      }
      if (callback) {
        callback(errFormatted);
      } else if (errFormatted) {
        if (this.errorHandler) {
          this.errorHandler(errFormatted);
        } else {
          // emit error ourselves on the socket for backwards compatibility
          this.socket.emit('error', errFormatted);
        }
      }
    };

    const buf = Buffer.from(message);
    try {
      this.messagesInFlight++;
      if (this.protocol === 'tcp') {
        this.socket.write(buf, 'ascii', handleCallback);
      } else if (this.protocol === 'unix_dgram') {
        this.socket.send(buf, handleCallback);
      } else {
        this.socket.send(buf, 0, buf.length, this.port, this.host, handleCallback);
      }
    } catch (err) {
      handleCallback(err);
    }
  }

  /**
   * Called every bufferFlushInterval to flush any buffer that is around
   */
  onBufferFlushInterval() {
    this.flushQueue();
  }

  /**
   * Close the underlying socket and stop listening for data on it.
   */
  close(callback) {
    // stop trying to flush the queue on an interval
    if (this.intervalHandle) {
      clearInterval(this.intervalHandle);
    }

    // flush the queue one last time, if needed
    this.flushQueue((err) => {
      if (err) {
        if (callback) {
          callback(err);
        }
        return;
      }

      // FIXME: we have entered callback hell, and this whole file is in need of an async rework

      // wait until there are no more messages in flight before really closing the socket
      let intervalAttempts = 0;
      const waitForMessages = setInterval(() => {
        intervalAttempts++;
        if (intervalAttempts > 10) {
          this.messagesInFlight = 0;
        }
        if (this.messagesInFlight <= 0) {
          clearInterval(waitForMessages);
          this._close(callback);
        }
      }, 50);
    });
  }

  /**
   * Really close the socket and handle any errors related to it
   */
  _close(callback) {
    // error function to use in callback and catch below
    let handledError = false;
    const handleErr = (err) => {
      const errMessage = `Error closing hot-shots socket: ${err}`;
      if (!handledError) {
        // The combination of catch and error can lead to some errors
        // showing up twice.  So we just show one of the errors that occur
        // on close.
        handledError = true;

        if (callback) {
          callback(new Error(errMessage));
        } else if (this.errorHandler) {
          this.errorHandler(new Error(errMessage));
        }
      }
    };

    if (this.errorHandler) {
      this.socket.removeListener('error', this.errorHandler);
    }

    // handle error and close events
    this.socket.on('error', handleErr);
    if (callback) {
      this.socket.on('close', err => {
        if (! handledError && callback) {
          callback(err);
        }
      });
    }

    try {
      if (this.protocol === 'tcp') {
        this.socket.destroy();
      } else {
        this.socket.close();
      }
    } catch (err) {
      handleErr(err);
    }
  }
}

module.exports = Transport;
